crates/core/src/factories/core_publish/test.rs

```diff
@@ -0,0 +1,129 @@
+};
+
+use crate::{
+    default_test_builder,
+    factories::{core_publish::CorePublish, MemoryOp},
+};
+
+use super::CorePublishConfig;
+
+async fn setup_test(
+    config: &CorePublishConfig,
+) -> (CorePublish, DynOpStore, Url) {
+    let builder =
+        Arc::new(default_test_builder().with_default_config().unwrap());
+
+    let op_store = builder
+        .op_store
+        .create(builder.clone(), TEST_SPACE_ID)
+        .await
+        .unwrap();
+
+    #[derive(Debug)]
+    struct NoopHandler;
+    impl TxBaseHandler for NoopHandler {}
+    impl TxHandler for NoopHandler {}
+    impl TxSpaceHandler for NoopHandler {}
+    impl SpaceHandler for NoopHandler {}
+
+    let transport = builder
+        .transport
+        .create(builder.clone(), Arc::new(NoopHandler))
+        .await
+        .unwrap();
+
+    let fetch = builder
+        .fetch
+        .create(
+            builder.clone(),
+            TEST_SPACE_ID,
+            op_store.clone(),
+            transport.clone(),
+        )
+        .await
+        .unwrap();
+
+    let url =
+        transport.register_space_handler(TEST_SPACE_ID, Arc::new(NoopHandler));
+
+    (
+        CorePublish::new(config.clone(), TEST_SPACE_ID, fetch, transport),
+        op_store,
+        url.unwrap(),
+    )
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn published_ops_can_be_retrieved() {
+    enable_tracing();
+
+    let (core_publish_1, op_store_1, _) =
+        setup_test(&CorePublishConfig::default()).await;
+    let (_core_publish2, op_store_2, url_2) =
+        setup_test(&CorePublishConfig::default()).await;
+
+    let incoming_op_1 = MemoryOp::new(Timestamp::now(), vec![1]);
+    let incoming_op_id_1 = incoming_op_1.compute_op_id();
+    let incoming_op_2 = MemoryOp::new(Timestamp::now(), vec![2]);
+    let incoming_op_id_2 = incoming_op_2.compute_op_id();
+
+    op_store_1
+        .process_incoming_ops(vec![incoming_op_1.into(), incoming_op_2.into()])
+        .await
+        .unwrap();
+
+    core_publish_1
+        .publish_ops(
+            vec![incoming_op_id_1.clone(), incoming_op_id_2.clone()],
+            url_2,
+        )
+        .await
+        .unwrap();
+
+    let ops = tokio::time::timeout(
+        std::time::Duration::from_millis(1000),
+        async move {
+            loop {
+                let ops = op_store_2
+                    .retrieve_ops(vec![
+                        incoming_op_id_1.clone(),
+                        incoming_op_id_2.clone(),
+                    ])
+                    .await
+                    .unwrap();
+
+                if ops.len() == 2 {
+                    return ops;
+                }
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+        },
+    )
+    .await
+    .unwrap();
+
+    assert!(ops.len() == 2);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn publish_to_invalid_url_does_not_error() {
```

Comment:
I understand the intention of the test, but it doesn't test that the outgoing publish task does not error, it tests that sending to that task doesn't error. It seems there's no way to test the publish doesn't error, unless you add some test only field.

Reply:
> it tests that sending to that task doesn't error

How do you arrive at this conclusion? I think what this test actually *is* missing indeed is a sleep so that there is time for the publish to occur in the receiving end of the channel. But then it does indeed test the publish itself I think as I'm seeing the following tracing error being logged:

```
could not send publish ops: Connection Closed (src: None) op_id=rhfL-NCni9XNqVteNtTNzxYxpIDy-js2kwDN_Wedhbo peer_url=ws://notanexistingurl:80
```

Reply:
I added a sleep now at the end of that test to make sure it makes it all the way to the publish step.

Reply:
Can we check in less than 3s? I'm on the fence about this. I do want us to check that sending doesn't error when we can't reach the other side and that does require waiting a bit but a fixed sleep will mean that we always have to wait 3s for this test to pass.

I have what I think might be a better suggestion that I didn't think of earlier. What about doing this "bad send" then following it with a valid send? That will then check that everything is still working after the thing that might have caused a fault?

Reply:
Sounds good to me. Would it be too cheecky to sneak that into the `published_ops_can_be_retrieved` test before sending to the valid url?

Reply:
`publish_ops` returns a future `K2Result` that always returns `Ok(())`.

In the function body the op ids are sent to the `outgoing_publish_ops_tx` channel. If that fails, the error is logged. If the receiving task then fails to send something, the error is again logged. But in no place is an error returned.

Reply:
Yes that's correct. And the test tests that things can still proceed smoothly in that case. Do you think the behavior should be different or that we should have another test that tests something else additionally?

Reply:
I renamed the test now and changed the logic to what @ThetaSinner proposed above (provided that I understood it correctly).